### PR TITLE
fix: removes extra domains from `SelectedNetworkController` domain state

### PR DESCRIPTION
**Pull request by adonesky1**
- Bumps to the the latest version of `@metamask/selected-network-controller` which fixes issue where the `SelectedNetworkController` adds any and all domains the user visits to its `domains` state whether or not the user has connected to these sites. 
- Adds a migration to clean up the undesired state caused by this bug.
- Adds logic to listen for updates to permissions state and add newly connected/permissioned dapps to `domains` state in the `SelectedNetworkController`. I've created a ticket [here](https://app.zenhub.com/workspaces/wallet-api-platform-63bee08a4e3b9d001108416e/issues/gh/metamask/metamask-planning/2142) to move this logic into the `SelectedNetworkController` itself.
- Adds logic to couple the `useRequestQueue` feature flag with the `perDomainNetwork` feature flag. I've created a ticket [here](https://app.zenhub.com/workspaces/wallet-api-platform-63bee08a4e3b9d001108416e/issues/gh/metamask/metamask-planning/2143) to remove the latter state since it is redundant and requires keeping these two booleans in sync.
- Removes calls to `setNetworkClientIdForMetamask` since the [latest version of `@metamask/selected-network-controller`](https://github.com/MetaMask/core/blob/main/packages/selected-network-controller/CHANGELOG.md#800) removes this method (and the `metamask` domain from the `domains` state entirely).
- Adds logic to use provider and blockTracker from the `SelectedNetworkController` when feature flag is on and connected origin is in `domains` state (which can only happen when the dapp is connected) in `setupProviderEngine` for RPC middleware stack.
- Fixes a bug where if a dapp prompts an `addEthereumChain` request and the user then accepts the subsequent `switchEthereumChain` call that is chained on, the `networkClientID` set for that dapp in the `SelectedNetworkController` domains state will not be updated.

Manual Testing Steps

well first you need to [comment out some code](https://github.com/MetaMask/metamask-extension/pull/22964) that is causing [a separate bug (found on develop)](https://github.com/MetaMask/metamask-extension/issues/22937) but one that will make manually testing this a nightmare.

1. Go to `settings` > `experimental` > `Select networks for each site` and toggle this setting on
2. Put the globally selected network to ethereum mainnet
3. Connect to [uniswap](https://app.uniswap.org/swap)
4. Connect to [pancakeswap](https://pancakeswap.finance/)
5. Vist [Aave](https://app.aave.com/) but don't connect the wallet
6. Open devtools in the wallet UI and run the follow command in the console:
```
const { metamask } = await window.stateHooks.getCleanAppState();
console.log(metamask.domains)
```
7. You should see the following log:
```
{https://app.uniswap.org: 'mainnet', https://pancakeswap.finance: 'mainnet'}
```
8. Now call `switchEthereumChain` on one of the connected dapps (they each have UI elements that do this)
9. Now go back to the other dapp (where the chain hasn't been switched) and initiate a transaction
10. You should receive a `switchEthereumChain` confirmation back to Ethereum Mainnet before the transaction confirmation is show.

### CODE REVIEW

File: app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js [hunk 1]
```diff
@@ -26,7 +26,6 @@ const switchEthereumChain = {
     setActiveNetwork: true,
     requestUserApproval: true,
     getNetworkConfigurations: true,
-    getNetworkClientIdForDomain: true,
```
Comment (adonesky1) [inline app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js:29]: this wasn't being used?

File: app/scripts/metamask-controller.js [hunk 2]
```diff
@@ -493,11 +494,31 @@ export default class MetamaskController extends EventEmitter {
...
+    // If the perDomainNetwork feature flag is enabled, use metamask origin provider and block tracker
+    // as the proxy passed to the controllers instantiated below, otherwise use the network controller's provider and block tracker
+    // let networkClient;
+    // if (this.selectedNetworkController.state.perDomainNetwork) {
+    //   networkClient =
+    //     this.selectedNetworkController.getProviderAndBlockTracker(
+    //       METAMASK_DOMAIN,
+    //     );
+    // } else {
+    // }
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:516]: for testing purposes due to https://github.com/MetaMask/metamask-extension/issues/22937 / https://github.com/MetaMask/metamask-extension/pull/22964

File: ui/components/ui/new-network-info/new-network-info.js [hunk 3]
```diff
@@ -42,29 +42,29 @@ export default function NewNetworkInfo() {
...
-  useEffect(() => {
-    checkTokenDetection();
-    // we want to only fetch once
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
+  // useEffect(() => {
+  //   checkTokenDetection();
+  //   // we want to only fetch once
+  //   // eslint-disable-next-line react-hooks/exhaustive-deps
+  // }, []);
```
Comment (adonesky1) [inline ui/components/ui/new-network-info/new-network-info.js:67]: for testing purposes due to https://github.com/MetaMask/metamask-extension/issues/22937 / https://github.com/MetaMask/metamask-extension/pull/22964

File: app/scripts/metamask-controller.js [hunk 4]
```diff
@@ -4913,10 +4940,6 @@ export default class MetamaskController extends EventEmitter {
             this.networkController,
           ),
         setActiveNetwork: (networkClientId) => {
-          // setActiveNetwork is used for custom (non-infura) networks, and for these networks networkClientId === networkConfigurationId
-          this.selectedNetworkController.setNetworkClientIdForMetamask(
-            networkClientId,
-          );
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:4919]: this is no longer necessary because `SelectedNetworkController` no longer keeps track of the globally `selectedNetworkClientId`

File: app/scripts/metamask-controller.js [hunk 5]
```diff
@@ -4938,8 +4961,6 @@ export default class MetamaskController extends EventEmitter {
         ),
         getProviderConfig: () => this.networkController.state.providerConfig,
         setProviderType: (type) => {
-          // when using this format, type happens to be the same as the networkClientId...
-          this.selectedNetworkController.setNetworkClientIdForMetamask(type);
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:4942]: this is no longer necessary because `SelectedNetworkController` no longer keeps track of the globally `selectedNetworkClientId`

**Comment by adonesky1** [comment]
@metamaskbot update-policies

File: app/scripts/metamask-controller.js [hunk 6]
```diff
@@ -2417,6 +2442,17 @@ export default class MetamaskController extends EventEmitter {
...
+          // When permissions are added for a given origin we need to
+          // set the network client id for that origin to the current
+          // globally selected network client id.
+          if (accounts.length > 0) {
+            this.selectedNetworkController.setNetworkClientIdForDomain(
+              origin,
+              this.networkController.state.selectedNetworkClientId,
+            );
+          } else {
+            // TODO remove domain from selectedNetworkController when permissions are removed?
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:2454]: I don't think theres any harm in leaving this state as if the user ever reconnects to this domain it will hit the logic right above and reset the selectedNetworkClientId to the globally selected network but probably cleaner to add a method to remove a domain from state on the SelectedNetworkController. wdyt @BelfordZ

File: app/scripts/metamask-controller.js [hunk 7]
```diff
@@ -493,11 +494,34 @@ export default class MetamaskController extends EventEmitter {
...
+    });
+
+    // couples the useRequestQueue featureflag with the perDomainNetwork feature flag
+    this.preferencesController.store.subscribe(({ useRequestQueue }) => {
+      if (
+        useRequestQueue !==
+        this.selectedNetworkController.state.perDomainNetwork
+      ) {
+        this.selectedNetworkController.setPerDomainNetwork(useRequestQueue);
+      }
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:518]: we should probably just remove this redundant `perDomainNetwork` state from the `SelectedNetworkController` and just query for the `useRequestQueue` state on the `PreferencesController` wherever we need it. But this will require another update to the `SelectedNetworkController` and should be done in a separate PR.

File: app/scripts/metamask-controller.js [hunk 8]
```diff
@@ -2417,6 +2441,17 @@ export default class MetamaskController extends EventEmitter {
...
         for (const [origin, accounts] of changedAccounts.entries()) {
           this._notifyAccountsChange(origin, accounts);
+          // When permissions are added for a given origin we need to
+          // set the network client id for that origin to the current
+          // globally selected network client id.
+          if (accounts.length > 0) {
+            this.selectedNetworkController.setNetworkClientIdForDomain(
+              origin,
+              this.networkController.state.selectedNetworkClientId,
+            );
```
Comment (jiexi) [inline app/scripts/metamask-controller.js:2451]: this is only necessary because we access domain state directly, right? because `getNetworkClientIdForDomain()` would already return the selectedNetworkClientId as a fallback

File: app/scripts/metamask-controller.js (hunk 6 above)
Comment (BelfordZ) [reply app/scripts/metamask-controller.js:2453]: yeah I think we are good to go without it, at least I cant think of any bug that would arise from leaving the state as it is.

File: app/scripts/lib/setupSentry.js [hunk 9]
```diff
@@ -225,7 +225,7 @@ export const SENTRY_BACKGROUND_STATE = {
     useTokenDetection: true,
     useRequestQueue: true,
   },
-  SelectedNetworkController: { domains: true, perDomainNetwork: false },
+  SelectedNetworkController: { domains: false, perDomainNetwork: true },
```
Comment (adonesky1) [inline app/scripts/lib/setupSentry.js:228]: We do not want to include domains in our sentry reports since it is potentially sensitive and violates our goal to preserve privacy, but we do want to include the perDomainNetwork toggle since it is not sensitive and potentially useful for diagnosing bugs.

File: app/scripts/metamask-controller.js (hunk 8 above)
Comment (adonesky1) [reply app/scripts/metamask-controller.js:2456]: So this is actually the point at which we want to add the newly connected dapp to the `domains` state in `SelectedNetworkController`. 

However, this is another candidate for something that can/should be pulled directly into the `SelectedNetworkController`

Comment (jiexi) [reply app/scripts/metamask-controller.js]: Sounds good. the less MetaMaskController has to be concerned about, the better

File: app/scripts/metamask-controller.js [hunk 10]
```diff
@@ -2422,6 +2446,19 @@ export default class MetamaskController extends EventEmitter {
...
+          if (accounts.length > 0) {
+            // TODO add this logic directly in the SelectedNetworkController
+            this.selectedNetworkController.setNetworkClientIdForDomain(
+              origin,
+              this.networkController.state.selectedNetworkClientId,
+            );
+          } else {
+            // TODO remove domain from selectedNetworkController when permissions are removed
+            // Add this logic directly in the SelectedNetworkController
+          }
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:2461]: There is a WIP PR against the `SelectedNetworkController` to pull this logic into that controller instead: https://github.com/MetaMask/core/pull/3969

**Comment by jiexi** [comment]
tested, working

File: app/scripts/metamask-controller.js [hunk 11]
```diff
@@ -520,6 +530,20 @@ export default class MetamaskController extends EventEmitter {
         ),
     });
 
+    // Couple the useRequestQueue featureflag with the perDomainNetwork feature flag
```
Comment (jiexi) [inline app/scripts/metamask-controller.js:533]:
```suggestion
    // Couple the useRequestQueue feature flag with the perDomainNetwork feature flag
```

nit

File: app/scripts/metamask-controller.js [hunk 12]
```diff
@@ -2609,7 +2646,7 @@ export default class MetamaskController extends EventEmitter {
       isUnlocked: this.isUnlocked(),
       accounts: await this.getPermittedAccounts(origin),
       ...this.getProviderNetworkState(
-        this.preferencesController.getUseRequestQueue() ? origin : undefined,
+        this.preferencesController?.getUseRequestQueue() ? origin : undefined,
```
Comment (jiexi) [inline app/scripts/metamask-controller.js:2649]: why is this needed now?

File: app/scripts/metamask-controller.js [hunk 13]
```diff
@@ -4728,25 +4760,21 @@ export default class MetamaskController extends EventEmitter {
...
-
-    let proxyProviderForDomain = provider;
-    if (this.preferencesController.getUseRequestQueue()) {
-      proxyProviderForDomain =
-        this.selectedNetworkController.getProviderAndBlockTracker(
-          origin,
-        ).provider;
+    // Use proxyClient from selectedNetworkController when:
+    // 1. feature flag for perDomainNetwork is on
+    // 2  the origin is in the selectedNetworkController's `domains` state
```
Comment (jiexi) [inline app/scripts/metamask-controller.js:4765]: I think it may be more helpful to future devs to explain why we prefer using the global as opposed to setting the networkClient for the origin if it doesn't exist already and then using that. I could see someone trying to set a default here and not having enough background/context to understand why we don't want to do that

File: test/e2e/webdriver/driver.js [hunk 14]
```diff
@@ -546,6 +546,43 @@ class Driver {
...
+        );
+
+        if (handleUrl === `${url}/`) {
+          return handle;
+        }
+      }
+      await this.delay(delayStep);
+      timeElapsed += delayStep;
+      // refresh the window handles
+      windowHandles = await this.driver.getAllWindowHandles();
```
Comment (jiexi) [inline test/e2e/webdriver/driver.js:580]: is it possible to do this all within the `await retry`?

```
await retry(
          {
            retries,
            delay: retryDelay,
          },
          async () => {
            windowHandles = await this.driver.getAllWindowHandles();
            for (const handle of windowHandles) {
              await this.driver.switchTo().window(handle);
              const handleUrl await this.driver.getCurrentUrl();
              if (handleUrl === `${url}/`) {
                return true;
              }
            }
          },
        );

```

Comment (jiexi) [reply test/e2e/webdriver/driver.js:580]: something like that?

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js [hunk 15]
```diff
@@ -294,6 +296,7 @@ async function addEthereumChainHandler(
         fromNetworkConfiguration: getProviderConfig(),
       },
     });
+    setNetworkClientIdForDomain(req.origin, networkConfigurationId);
```
Comment (adonesky1) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:299]: This fixes a bug where if a dapp prompts an add-ethereum-chain request and the user then accepts the subsequent `switchEthereumChain` call that is chained on, the `networkClientID` set for that dapp in the `SelectedNetworkController` domains state will not be updated.

**Comment by darkwing** [comment]
I can confirm this PR fixes my `domains` key/value issue!

File: app/scripts/metamask-controller.js [hunk 16]
```diff
@@ -2620,9 +2673,12 @@ export default class MetamaskController extends EventEmitter {
    * @param {string} origin - The origin identifier for which network state is requested (default: 'metamask').
    * @returns {object} An object containing important network state properties, including chainId and networkVersion.
    */
-  getProviderNetworkState(origin = 'metamask') {
+  getProviderNetworkState(origin = METAMASK_DOMAIN) {
     let chainId;
-    if (this.preferencesController.getUseRequestQueue()) {
+    if (
+      this.preferencesController?.getUseRequestQueue() &&
```
Comment (shanejonas) [inline app/scripts/metamask-controller.js:2679]:
```suggestion
      this.preferencesController.getUseRequestQueue() &&
```
`this.preferencesController` should always exist

Comment (adonesky1) [reply app/scripts/metamask-controller.js:2679]: unless, as appears to be the case here in some scenarios, this method is called before the controller is instantiated.

File: test/e2e/webdriver/driver.js (hunk 14 above)
Comment (adonesky1) [reply test/e2e/webdriver/driver.js:580]: cc @tmashuang ☝️

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js [hunk 17]
```diff
@@ -294,6 +298,9 @@ async function addEthereumChainHandler(
         fromNetworkConfiguration: getProviderConfig(),
       },
     });
+    if (hasPermission(req.origin)) {
+      setNetworkClientIdForDomain(req.origin, networkConfigurationId);
```
Comment (Gudahtt) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:302]: This solution seems OK, but it makes me wonder if similar cases exist elsewhere.

Have we considered listening to permission state changes, and automatically setting a selected chain when a new origin is granted permissions? That would ensure that we never "forget" to set this initial selected network client.

Comment (adonesky1) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:302]: > Have we considered listening to permission state changes, and automatically setting a selected chain when a new origin is granted permissions

Yes indeed. In fact we're doing that in this PR. See [here](https://github.com/MetaMask/metamask-extension/pull/23130/files#r1503052895).

The problem here is not the **initial** network being set, its that we need to change the networkClientId for a dapp that already is permissioned.

Comment (Gudahtt) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:302]: Ahh I see,  I follow now, thanks

File: app/scripts/metamask-controller.js (hunk 7 above)
Comment (adonesky1) [reply app/scripts/metamask-controller.js]: This dedupe is happening here btw: https://github.com/MetaMask/core/pull/3989

File: app/scripts/metamask-controller.js (hunk 16 above)
Comment (adonesky1) [reply app/scripts/metamask-controller.js]: not seeing this issue anymore so removing this for now: https://github.com/MetaMask/metamask-extension/pull/23130/commits/f2c2dbdb049f2e7268be45d684c8714acc1833f9

File: app/scripts/metamask-controller.js (hunk 13 above)
Comment (adonesky1) [reply app/scripts/metamask-controller.js]: Chained the slant of the comments here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/f2c2dbdb049f2e7268be45d684c8714acc1833f9 . Let me know what you think

File: app/scripts/metamask-controller.js (hunk 12 above)
Comment (adonesky1) [reply app/scripts/metamask-controller.js]: not seeing this issue anymore so removing this for now: https://github.com/MetaMask/metamask-extension/commit/f2c2dbdb049f2e7268be45d684c8714acc1833f9

File: test/e2e/webdriver/driver.js (hunk 14 above)
Comment (tmashuang) [reply test/e2e/webdriver/driver.js:580]: I tried what @jiexi proposed and It didn't seem to work. Looks like it wasn't able to break out of the `retry()`

Comment (jiexi) [reply test/e2e/webdriver/driver.js:580]: ahhh, i'm sorry. Thank you for taking on the burden of trying it. Apologies 🙏

File: test/e2e/webdriver/driver.js [hunk 18]
```diff
@@ -546,6 +546,43 @@ class Driver {
...
+          async () => {
+            await this.driver.switchTo().window(handle);
+            return await this.driver.getCurrentUrl();
+          },
+        );
+
+        if (handleUrl === `${url}/`) {
+          return handle;
+        }
+      }
```
Comment (jiexi) [inline test/e2e/webdriver/driver.js:576]: ahh i'm sorry to comment on this one again. It just seems weird to have what is essentially two retry/timeout loops. Is this inner `retry` necessary? 

timeElapsed isn't really the time elapsed since it doesn't account for how long the execution of the loop took

Comment (tmashuang) [reply test/e2e/webdriver/driver.js:576]: I believe the retry is necessary, maybe the `while (timeElapsed <= timeout)` part might not be though. These methods have been combined to combat the flaky tests [PR](https://github.com/MetaMask/metamask-extension/pull/21394) and internal discussions as to the best approach [internal disc](https://consensys.slack.com/archives/GTQAGKY5V/p1697223609357109?thread_ts=1697150920.731179&cid=GTQAGKY5V)  and an open issue https://github.com/MetaMask/metamask-extension/issues/21403 for future refactoring in the future.

Comment (tmashuang) [reply test/e2e/webdriver/driver.js:576]: Thinking more largely, we could combine all the switching windows methods together `switchToWindow`, `switchToWindowWithTitle`, and `switchToWindowWithUrl`.

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js (hunk 17 above)
Comment (adonesky1) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:302]: I wonder if we can also add this directly in the `SelectedNetworkController` by listening for accepted `switchEthereumChain` requests from the `ApprovalController`? wdyt @Gudahtt @mcmire @BelfordZ

Comment (adonesky1) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:302]: Looked into this doesn't look very straightforward unfortunately.

File: test/e2e/webdriver/driver.js (hunk 18 above)
Comment (jiexi) [reply test/e2e/webdriver/driver.js:576]: Thank you for explaining 🙏

**Review by jiexi — DISMISSED** [review]

File: app/scripts/metamask-controller.js [hunk 19]
```diff
@@ -2618,9 +2671,12 @@ export default class MetamaskController extends EventEmitter {
    * @param {string} origin - The origin identifier for which network state is requested (default: 'metamask').
    * @returns {object} An object containing important network state properties, including chainId and networkVersion.
    */
-  getProviderNetworkState(origin = 'metamask') {
+  getProviderNetworkState(origin = METAMASK_DOMAIN) {
     let chainId;
-    if (this.preferencesController.getUseRequestQueue()) {
+    if (
```
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:2676]: Just noticed a bug in this function: the `networkVersion` returned is always for the globally selected network, even when this feature is on. That would result in the provider having a `networkId` that contradicts the `chainId`. Not good.

Not a blocker for this PR since it wasn't introduced here, but we should fix this as well.

File: app/scripts/metamask-controller.js [hunk 20]
```diff
@@ -5767,7 +5809,7 @@ export default class MetamaskController extends EventEmitter {
   }
 
   _notifyChainChange() {
-    if (this.preferencesController.getUseRequestQueue()) {
+    if (this.preferencesController?.getUseRequestQueue()) {
```
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:5812]: Do we still need this change?

Maybe similar to [this case](https://github.com/MetaMask/metamask-extension/pull/23130#discussion_r1504540531), in that it's no longer needed

File: app/scripts/migrations/111.ts [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  transformState(versionedData.data);
+  return versionedData;
+}
+
+function transformState(state: Record<string, any>) {
+  if (
+    !hasProperty(state, 'SelectedNetworkController') ||
+    !isObject(state.SelectedNetworkController) ||
```
Comment (Gudahtt) [inline app/scripts/migrations/111.ts:34]: Nit: In other migrations we distinguished between the "Missing state that is expected to be missing" scenarios, and "Malformed/corrupted state".

We would expect the `SelectedNetworkController` state to be missing if the user upgraded from an older wallet, so returning `state` unchanged for that scenario seems good. But if that state exists, the remaining three conditions should be true for all users unless the state was somehow corrupted. So we could return an error in those scenarios rather than returning state unchanged, to alert us to that state corruption.

File: app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js
Comment (Gudahtt) [inline app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js:1]: I'm seeing this change entry in v8 of the SelectedNetworkController:

>`setNetworkClientIdForDomain` now fails and throws an error if the passed in `domain` is not currently permissioned in the `PermissionsController`

yet this middleware will call `setNetworkClientIdForDomain` irrespective of whether the domain has permissions or not

File: test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js [hunk 22]
```diff
@@ -23,8 +25,8 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
         dapp: true,
         fixtures: new FixtureBuilder()
           .withNetworkControllerDoubleGanache()
-          .withPermissionControllerConnectedToTwoTestDapps()
-          .withSelectedNetworkControllerPerDomain()
+          // .withPermissionControllerConnectedToTwoTestDapps()
+          // .withSelectedNetworkControllerPerDomain() // TODO: Add back when refreshing state doesn't revert per dapp selected network. Remove having to connect to dapp
```
Comment (Gudahtt) [inline test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:29]: Hmm, what does this comment refer to? I'm not familiar with this problem

File: app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js
Comment (adonesky1) [reply app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js:1]: Good call! This is not a flow I thought of, but is technically possible as well: where a dapp does not connect but prompts the wallet to switch chains. Will add permission check here as well!

File: app/scripts/metamask-controller.js (hunk 20 above)
Comment (adonesky1) [reply app/scripts/metamask-controller.js]: Reverted here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/9af40f307ef4dcb2536aee62a5aef6ebb3126f6e

File: app/scripts/metamask-controller.js [hunk 23]
```diff
@@ -4726,25 +4774,25 @@ export default class MetamaskController extends EventEmitter {
...
+        this.selectedNetworkController.getProviderAndBlockTracker(origin);
+    } else {
+      // if perDomainNetwork is false we want to use the globally selected network provider/blockTracker
+      // since this means the per domain network feature is disabled
+
+      // if the origin is not in the selectedNetworkController's `domains` state,
+      // this means that origin does not have permissions (is not connected to the wallet)
+      // and will therefore not have its own selected network even if perDomainNetwork is true
+      // and so in this case too we want to use the globally selected network provider/blockTracker
+      proxyClient = {
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/metamask-controller.js]: Done here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/9af40f307ef4dcb2536aee62a5aef6ebb3126f6e

File: app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js
Comment (adonesky1) [reply app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js:1]: added here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/7cbb194fe92181d0c2bfe8eb9f2be45de4d54ef3

File: app/scripts/metamask-controller.js [hunk 24]
```diff
@@ -492,11 +493,20 @@ export default class MetamaskController extends EventEmitter {
...
+          `PermissionController:hasPermissions`,
+        ],
         allowedEvents: ['NetworkController:stateChange'],
       }),
+      state: initState.SelectedNetworkController,
     });
 
+    const networkClient = this.networkController.getProviderAndBlockTracker();
+    this.provider = networkClient.provider;
+    this.blockTracker = networkClient.blockTracker;
```
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:508]: Following 9af40f307ef4dcb2536aee62a5aef6ebb3126f6e, this seems to no longer be used

```suggestion
```

File: ui/components/ui/new-network-info/new-network-info.js [hunk 25]
```diff
@@ -61,7 +61,7 @@ export default function NewNetworkInfo() {
   }, [providerConfig.chainId]);
 
   useEffect(() => {
-    checkTokenDetection();
+    // checkTokenDetection();
```
Comment (Gudahtt) [inline ui/components/ui/new-network-info/new-network-info.js:64]: Was this change intentional?

**Comment by tmashuang** [comment]
Some issues I ran into with my testing of the setting would be that there wasn't a good way to clear all the domains when debugging. It can be cleared individually via the connected dapp modal, but I would hope that the toggle would clear the `domain` state when turned off. Reiterating the issue with state refreshing not persisting the network per domain again as every update to the extension will revert the behavior. Also, the persisting `eth_chainId` sent back to the dapp even though the setting is toggled on.

File: ui/components/ui/new-network-info/new-network-info.js (hunk 25 above)
Comment (adonesky1) [reply ui/components/ui/new-network-info/new-network-info.js]: nope! reverted here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/0e8ee1c6417c0d4bbb0be7222bd03e27d5a5d7ab

File: app/scripts/migrations/111.ts (hunk 21 above)
Comment (adonesky1) [reply app/scripts/migrations/111.ts]: added error logging here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/bae4ca69e0077ef03f7c781c066cd31d3698e7e0

File: app/scripts/metamask-controller.js (hunk 24 above)
Comment (adonesky1) [reply app/scripts/metamask-controller.js]: removed here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/bf5bc3ed7b7f769934f626e0752342261f87ba70

File: app/scripts/metamask-controller.js (hunk 19 above)
Comment (adonesky1) [reply app/scripts/metamask-controller.js:2672]: Heres a ticket: https://github.com/MetaMask/MetaMask-planning/issues/2176

File: test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js (hunk 22 above)
Comment (adonesky1) [reply test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:29]: I'm not totally understanding this either. Currently working with Thomas to figure it out!

Comment (adonesky1) [reply test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js]: ok this was a misunderstanding. I worked through it with @tmashuang  and we've updated the e2e tests here https://github.com/MetaMask/metamask-extension/pull/23130/commits/85a1d28989f7f61472c0ad65a787a98023957752 so that they work as expected now. Let me know if you have any further questions @Gudahtt

File: test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js [hunk 26]
```diff
@@ -85,25 +85,48 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
         // Open Dapp Two
         await openDapp(driver, undefined, DAPP_ONE_URL);
 
-        // Window Handling
-        const windowHandles = await driver.getAllWindowHandles();
-        const dappOne = windowHandles[1];
-        const dappTwo = windowHandles[2];
+        // // Connect to dapp
```
Comment (tmashuang) [inline test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:88]:
```suggestion
        // Connect to dapp
```

File: test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js [hunk 27]
```diff
@@ -40,34 +42,32 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
...
 
         // Open Dapp One
         await openDapp(driver, undefined, DAPP_URL);
 
+        // Connect to dapp
+        await driver.clickElement('#connectButton');
+
+        await driver.delay(regularDelayMs);
+
+        // // Connect to Dapp
```
Comment (tmashuang) [inline test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:56]:
```suggestion
        // Connect to Dapp
```

**Comment by Gudahtt** [comment]
The diff looks good! I can re-review when CI passes. It looks like at least one of the CI failures is related to these changes, the e2e test failure.

File: test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js [hunk 28]
```diff
@@ -40,34 +42,32 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
...
 
         // Open Dapp One
         await openDapp(driver, undefined, DAPP_URL);
 
+        // Connect to dapp
+        await driver.clickElement('#connectButton');
+
+        await driver.delay(regularDelayMs);
+
+        // Connect to Dapp
```
Comment (jiexi) [inline test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:56]:
```suggestion
```

this comment already exists above on line 51

File: test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js [hunk 29]
```diff
@@ -85,25 +85,48 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
...
-        // Window Handling
-        const windowHandles = await driver.getAllWindowHandles();
-        const dappOne = windowHandles[1];
-        const dappTwo = windowHandles[2];
+        // Connect to dapp
+        await driver.clickElement('#connectButton');
+
+        await driver.delay(regularDelayMs);
+
+        // Connect to Dapp
```
Comment (jiexi) [inline test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:93]:
```suggestion
```

**Review by jiexi — DISMISSED** [review]

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Review by jiexi — APPROVED** [review]
